numba/dispatcher.py

```diff
@@ -135,6 +134,28 @@ def _get_implementation(self, args, kws):
     '_CompileStats', ('cache_path', 'cache_hits', 'cache_misses'))
 
 
+class _CompilingCounter(object):
+    """
+    A simple counter that increment in __enter__ and decrement in __exit__.
+    """
+
+    def __init__(self):
+        self.counter = 0
+
+    def __enter__(self):
+        assert self.counter >= 0
+        self.counter += 1
```

Comment:
Is this threadsafe?  Unless the load, increment, and save can be done in a single opcode, I don't think the GIL will protect this from a race condition.